### ISSUE TITLE
Project Structure

### ISSUE THREAD

**Issue by eagerestwolf** [neutralinojs/neutralinojs#380]
So, there are a few niggles I have about the project structure, and app creation process using the CLI.

First, the CLI, without asking the user (like Vue and React do) defaults to `npm` as the package installer (side note: does anyone actually still use npm?). This wouldn't be an issue, except `pnpm` and `yarn` now won't work and will throw errors until the `node_modules` folder and `package-lock.json` are both deleted, thus requiring that all of the packages are installed a second time. This likely only requires a slight tweak to the cli, such as an option. Alternatively, you could take the Vue/React approach and ask the user the first time which package manager they prefer and save that to a global config.

Second, and I get this is rather pedantic and nitpicky, but the project structure is very non-standard. The `neutralino` configs should probably be in their own specific area (say a `.neutralinorc` file in the project root a la ESLint, Prettier, and TypeScript). An example potential for a `.neutralinorc` file is included below this paragraph. The `app` directory would probably better be served as a `public` directory, similar to how Vue, React, Proton and Electron do it.

```jsonc
{
  // Provide defaults here
  "name": "hello-world",
  "port": 8080, // <- note the port number
  "url": "/",
  "modes": {
    "window": {
      "port": 0, // <- just override it here
      "title": "Hello World",
      "width": 640,
      "height": 480,
      "fullscreen": false,
      "alwaysOnTop": false,
      "iconFile": "public/icon.png",
      "enableInspector": false,
      "borderlessWindow": false
    },
    "cloud": {
      "blacklist": [ "os.runCommand" ]
    },
    "browser": {}
  }
}
```

Third, having raw executables in the repository isn't really a good idea, for obvious reasons. It would probably be a better idea to have a separate package, say `@neutralinojs/exec` that installs the executables to the `node_modules/.bin` folder via the `bin` property in `package.json`. This way the user can create an npm script for launching the app for testing and debugging. Also, while I'm on the topic of things in the root of the repository, there is 2 things that should probably be changed. The first is simple, just move `neutralino.png` to either the `app` dir or the proposed `public` dir. The other is also really simple, neutralino shouldn't create a log file unless there was an error, kind of like npm and yarn.

Finally, there's some minor template issues (probably just forgot to change these). There is a `funding.yml` file in the `.github` directory. That probably shouldn't be there, especially because it already has someone else's funding info filled in. Next, the `package.json` seems to allude to the fact that the template uses TypeScript (the project name is `neutralinojs-typescript`, but it doesn't). Finally, on the topic of the `package.json`, why not take the project name provided by the CLI and set that to the project name in the `package.json`

All in all, please don't think I'm hating on your project. This is actually a really interesting project, that shows a lot of promise. I'm just providing some constructive criticism in hopes of making it even better, and I'm willing to put my money where my mouth is. I'll fork this repo, check out the configuration stuffs and see if I can implement the `.neutralinorc` support, and pull it back. I won't touch the existing code for the time being, so that way existing projects don't break if you decide to keep it.

**Specifications**  
* NeutralinoJs CLI version: 2.3.3
* Platform: macOS 11 (arm64) [MacBook Pro 13" mid-2020]

**Comment by shalithasuranga** [comment]
Thank you very much for your suggestions. This is initially started as a kind of POC project. So, haven't thought about this stuff a lot earlier. Now we are working for a v2 release with your ideas https://github.com/neutralinojs/neutralinojs/milestone/11

**Comment by eagerestwolf** [comment]
@shalithasuranga Well, I'm glad you like the suggestions I put forward. I do completely understand how proof of concepts can sometimes become much more than planned. If you would like, I would definitely like to help with the v2 implementation, but I'm not a very efficient C++ developer. I took a look at the CLI project before and it seemed kind of odd to me how it's structured (I don't mean that in a bad way, every developer has their own unique style). Since v2 is coming and a major redesign could be accommodated since there will already be breaking changes, would you be open to the idea of a complete rewrite of the CLI (using TypeScript for type safety and underscore for efficiency)? I can probably have something up and running in a week or two for review (sorry, I work a full-time job during the week, so I only really can code on the weekends).

**Comment by shalithasuranga** [comment]
Hey really appreciate your willingness to contribute. The goal bind v2 is to make Neutralino production-ready and also make source code maintainable as explained in the [roadmap 2021](https://factsoverflow.com/neutralinojs-2021-roadmap). So the most critical things are,

- Getting macOS version synched with the main version.
- Error code system. 
- Making the project structure consistent in all templates.
- Implementing Vue and Angular starter kits.

Right now we didn't fee any maintainability issues with CLI. I think we may think about CLI improvements a bit later after v2. Sounds like you could help us to accomplish the 3rd and 4th items from the above list because you already suggested a great overview for the config file.

**Comment by eagerestwolf** [comment]
Absolutely, sorry for the radio silence by the way, just busy with work is all. I can look into the Vue and Angular kits over the course of the rest of this week and the weekend. I should be able to get Vue working at the very least (although the Vue CLI can sometimes be a stinker).

**Comment by shalithasuranga** [comment]
Hey.. I will close this one since v2 already has a nice structure based on this idea.We decided to offer one minimal template instead of many templates (hard to maintain). We are planning to write some how-to articles to show how to use frontend frameworks with Neutralino. See more: https://neutralino.js.org/docs/developer-environment/project-structure

Hope you will try Neutralinojs with Vue and Angular.. maybe you will write a blog post about those.
